drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2126 @@
+              /* Skip delete entry as it is already the
+               * last one.
+               */
+
+              return 0;
+            }
+        }
+      else if (pdata->len == wlk_ate.len &&
+               wlk_ate.expired == fs->erasestate)
+        {
+          /* Do not try to compare if lengths are not equal
+           * or prev one is deleted.
+           * Compare the data and if equal return 0.
+           */
+
+          rd_addr += wlk_ate.offset + wlk_ate.key_len;
+          rc = nvs_flash_block_cmp(fs, rd_addr, pdata->configdata,
+                                   pdata->len);
+          if (rc <= 0)
+            {
+              return rc;
+            }
+        }
+    }
+  else
+    {
+      /* Skip delete entry for non-existing entry. */
+
+      if (pdata->len == 0)
+        {
+          return 0;
+        }
+    }
+
+  /* Calculate required space if the entry contains data. */
+
+  if (data_size)
+    {
+      /* Leave space for gc_done ate */
+
+      required_space = data_size + sizeof(struct nvs_ate);
+    }
+
+  if (pdata->len == 0)
+    {
+      DEBUGASSERT(prev_found);
+
+      /* Comes into this branch only when a prev ate is found
+       * and it is not a delete entry.
+       * So lets delete by expiring the prev ate.
+       */
+
+      rc = nvs_expire_ate(fs, hist_addr);
+      if (rc < 0)
+        {
+          ferr("expire ate failed, addr %" PRIx32 "\n", hist_addr);
+          return rc;
+        }
+
+      /* Delete now requires no extra space, so skip write and gc. */
+
+      finfo("nvs_delete success\n");
+
+      return 0;
+    }
+
+  gc_count = 0;
+  block_to_write_befor_gc = fs->ate_wra >> ADDR_BLOCK_SHIFT;
+  while (1)
+    {
+      if (gc_count == fs->geo.neraseblocks)
+        {
+          /* Gc'ed all blocks, no extra space will be created
+           * by extra gc.
+           */
+
+          return -ENOSPC;
+        }
+
+      if (fs->ate_wra >= fs->data_wra + required_space)
+        {
+          finfo("Write entry, ate_wra=0x%" PRIx32 ", "
+                "data_wra=0x%" PRIx32 "\n",
+                fs->ate_wra, fs->data_wra);
+          rc = nvs_flash_wrt_entry(fs, hash_id, key, key_size,
+                                   pdata->configdata, pdata->len);
+          if (rc)
+            {
+              fwarn("Write entry failed\n");
+              return rc;
+            }
+
+          finfo("Write entry success\n");
+
+          /* Nvs is changed after gc, we will look for the second latest ate
+           * and expire it.
+           * After this operation, only the latest ate is valid.
+           * Expire the old one only if it is not deleted before(Or it is
+           * already expired)
+           */
+
+          if (prev_found && wlk_ate.expired == fs->erasestate)
+            {
+              finfo("prev entry exists, expire it\n");
+
+              /* If gc touched second latest ate, search for it again */
+
+              if (gc_count >= fs->geo.neraseblocks - 1 -
```

Comment:
why not directly expire the old entry

Reply:
The old entry will be moved after several gc. So in such case we cannot directly expire it (it is not there anymore), we have to search for the old one again.

Reply:
But, it shouldn't have more than one duplicated entry?

Reply:
there are more than one duplicated entries, when we are update an existing KV.
It happens after we finished writing the new KV, and right before we expired the old KV.
At that time, the old KV may be moved to someplaces else due to gc.

And We expire the old one only after we finished writing the new KV. Expire old kv must be after writiing the new KV, so that if power-loss happens when we haven't finished writing the new KV, the old kv is still valid.

Reply:
Optimize the nvs_read_hist， remove cnt.
Now the work flow will be gc, search for prev, write, expire prev.